README.md
# OpenConcept - A conceptual design toolkit with efficient gradients implemented in the OpenMDAO framework

### Authors: Benjamin J. Brelje and Eytan J. Adler

[![Build](https://github.com/mdolab/openconcept/workflows/Build/badge.svg?branch=main)](https://github.com/mdolab/openconcept/actions?query=branch%3Amain)
[![Coverage](https://codecov.io/gh/mdolab/openconcept/branch/main/graph/badge.svg?token=RR8CN3IOSL)](https://codecov.io/gh/mdolab/openconcept)
[![Documentation](https://readthedocs.com/projects/mdolab-openconcept/badge/?version=latest)](https://mdolab-openconcept.readthedocs-hosted.com/en/latest/?badge=latest)
[![PyPI](https://img.shields.io/pypi/v/openconcept)](https://pypi.org/project/openconcept/)
[![PyPI - Downloads](https://img.shields.io/pypi/dm/openconcept)](https://pypi.org/project/openconcept/)

OpenConcept is a new toolkit for the conceptual design of aircraft. OpenConcept was developed in order to model and optimize aircraft with electric propulsion at low computational cost. The tools are built on top of NASA Glenn's [OpenMDAO](http://openmdao.org/) framework, which in turn is written in Python.

OpenConcept is capable of modeling a wide range of propulsion systems, including detailed thermal management systems.
The following figure (from [this paper](https://doi.org/10.3390/aerospace9050243)) shows one such system that is modeled in the `N3_HybridSingleAisle_Refrig.py` example.

<h2 align="center">
    <img src="https://raw.githubusercontent.com/mdolab/openconcept/main/doc/_static/images/full_parallel_system_chiller.png" width="500" />
</h2>

The following charts show more than 250 individually optimized hybrid-electric light twin aircraft (similar to a King Air C90GT). Optimizing hundreds of configurations can be done in a couple of hours on a standard laptop computer.

![Example charts](https://raw.githubusercontent.com/mdolab/openconcept/main/doc/_static/images/readme_charts.png)

The reason for OpenConcept's efficiency is the analytic derivatives built into each analysis routine and component. Accurate, efficient derivatives enable the use of Newton nonlinear equation solutions and gradient-based optimization at low computational cost.

## Documentation

Automatically-generated documentation is available at (https://mdolab-openconcept.readthedocs-hosted.com/en/latest/).

To build the docs locally, install the `sphinx_mdolab_theme` via `pip`. Then enter the `doc` folder in the root directory and run `make html`. The built documentation can be viewed by opening `_build/html/index.html`. OpenAeroStruct is required (also installable via `pip`) to build the OpenAeroStruct portion of the source docs.

## Getting Started

OpenConcept can be pip installed directly from PyPI

```shell
pip install openconcept
```

To run the examples or edit the source code:
1. Clone the repo to disk (`git clone https://github.com/mdolab/openconcept`)
2. Navigate to the root `openconcept` folder
3. Run `pip install -e .` to install the package (the `-e` can be omitted if not editing the source)

Get started by following the tutorials in the documentation to learn the most important parts of OpenConcept.
The features section of the documentation describes most of the components and system models available in OpenConcept.

### Requirements

<!-- Remember to change doc/index.rst too! -->

OpenConcept is tested regularly on builds with the oldest and latest supported package versions. The package versions in the oldest and latest builds are the following:

| Package | Oldest | Latest |
| ------- | ------- | ------ |
| Python | 3.8 | 3.11 |
| OpenMDAO | 3.21 | latest |
| NumPy | 1.20 | 1.26 |
| SciPy | 1.7.0 | latest |
| OpenAeroStruct | 2.7.1 | 2.7.1 |

## Citation

Please cite this software by reference to the [conference paper](https://www.researchgate.net/publication/326263660_Development_of_a_Conceptual_Design_Model_for_Aircraft_Electric_Propulsion_with_Efficient_Gradients):

Benjamin J. Brelje and Joaquim R. R. A. Martins, "Development of a Conceptual Design Model for Aircraft Electric Propulsion with Efficient Gradients", 2018 AIAA/IEEE Electric Aircraft Technologies Symposium, AIAA Propulsion and Energy Forum, (AIAA 2018-4979) DOI: 10.2514/6.2018-4979

```
@inproceedings{Brelje2018a,
	address = {{C}incinnati,~{OH}},
	author = {Benjamin J. Brelje and Joaquim R. R. A. Martins},
	booktitle = {Proceedings of the AIAA/IEEE Electric Aircraft Technologies Symposium},
	doi = {10.2514/6.2018-4979},
	month = {July},
	title = {Development of a Conceptual Design Model for Aircraft Electric Propulsion with Efficient Gradients},
	year = {2018}
}
```

If using the integrated OpenAeroStruct VLM or aerostructural aerodynamic models, please cite the following [conference paper](https://www.researchgate.net/publication/357559489_Aerostructural_wing_design_optimization_considering_full_mission_analysis):

Eytan J. Adler and Joaquim R. R. A. Martins, "Efficient Aerostructural Wing Optimization Considering Mission Analysis", Journal of Aircraft, 2022. DOI: 10.2514/1.c037096

```
@article{Adler2022d,
	author = {Adler, Eytan J. and Martins, Joaquim R. R. A.},
	doi = {10.2514/1.c037096},
	issn = {1533-3868},
	journal = {Journal of Aircraft},
	month = {December},
	publisher = {American Institute of Aeronautics and Astronautics},
	title = {Efficient Aerostructural Wing Optimization Considering Mission Analysis},
	year = {2022}
}
```


changes.md
# Changes to make before merging

- [ ] Merge Shugo's work
- [x] Add docs:
  - [x] Examples
    - [x] Do simple example
    - [x] Show how aircraft model is made and how variables are passed
    - [x] Show analysis group
    - [x]  Show running the thing
  - [x] Describe features
    - [x] Aerodynamics
    - [x] Atmospherics
    - [x] Costs
    - [x] Mission analysis
    - [x] Propulsion
    - [x] Thermal
    - [x] Utilities
    - [x] Weights
  - [x] Maybe a list of papers that use OpenConcept like OpenMDAO does?
- [x] Add linting and formatting
- [ ] Fix TODO on first line of SteadyFlightPhase in solver_phases.py (ehh, it's one of the fun features of OpenConcept)
- [x] Remove old docs (old "features")

# Refactoring directory tree

- [x] Refactoring is done!

- openconcept
  - aerodynamics
  - atmospherics
  - costs
  - mission analysis
  - propulsion
    - thrust?
    - energy storage
      - batteries, (hydrogen in the future), etc.
    - layouts/architectures
  - thermal
  - utilities
    - math
  - weights

## Other refactoring

- [x] move examples folder to within openconcept directory! Will need to change the paths in the tutorial docs
- [ ] remove old code? For example, `OldIntegrator`


examples/HybridTwin.py
from __future__ import division
import sys
import os
import logging
import numpy as np
sys.path.insert(0, os.getcwd())
from openmdao.api import Problem, Group, ScipyOptimizeDriver
from openmdao.api import BalanceComp, ExplicitComponent, ExecComp, SqliteRecorder
from openmdao.api import DirectSolver, IndepVarComp, NewtonSolver, BoundsEnforceLS

# imports for the airplane model itself
from openconcept.analysis.aerodynamics import PolarDrag
from openconcept.utilities.math import AddSubtractComp
from openconcept.utilities.math.integrals import Integrator
from openconcept.utilities.dvlabel import DVLabel
from examples.methods.weights_twin_hybrid import TwinSeriesHybridEmptyWeight
from examples.propulsion_layouts.simple_series_hybrid import TwinSeriesHybridElectricPropulsionSystem
from examples.methods.costs_commuter import OperatingCost
from openconcept.utilities.dict_indepvarcomp import DictIndepVarComp
from examples.aircraft_data.KingAirC90GT import data as acdata
from openconcept.analysis.performance.mission_profiles import FullMissionAnalysis
from openconcept.utilities.linearinterp import LinearInterpolator
from openconcept.utilities.visualization import plot_trajectory

class AugmentedFBObjective(ExplicitComponent):
    def setup(self):
        self.add_input('fuel_burn', units='kg')
        self.add_input('ac|weights|MTOW', units='kg')
        self.add_output('mixed_objective', units='kg')
        self.declare_partials(['mixed_objective'], ['fuel_burn'], val=1)
        self.declare_partials(['mixed_objective'], ['ac|weights|MTOW'], val=1/100)
    def compute(self, inputs, outputs):
        outputs['mixed_objective'] = inputs['fuel_burn'] + inputs['ac|weights|MTOW']/100

class SeriesHybridTwinModel(Group):
    """
    A custom model specific to a series hybrid twin turboprop-class airplane
    This class will be passed in to the mission analysis code.

    """
    def initialize(self):
        self.options.declare('num_nodes', default=1)
        self.options.declare('flight_phase', default=None)

    def setup(self):
        nn = self.options['num_nodes']
        flight_phase = self.options['flight_phase']

        # any control variables other than throttle and braking need to be defined here
        controls = self.add_subsystem('controls', IndepVarComp(), promotes_outputs=['*'])
        controls.add_output('proprpm',val=np.ones((nn,))*2000, units='rpm')

        # assume TO happens on battery backup
        if flight_phase in ['climb', 'cruise','descent']:
            controls.add_output('hybridization',val=0.0)
        else:
            controls.add_output('hybridization',val=1.0)

        hybrid_factor = self.add_subsystem('hybrid_factor', LinearInterpolator(num_nodes=nn),
                                           promotes_inputs=[('start_val', 'hybridization'),
                                                            ('end_val', 'hybridization')])

        propulsion_promotes_outputs = ['fuel_flow','thrust']
        propulsion_promotes_inputs = ["fltcond|*", "ac|propulsion|*", "throttle", "propulsor_active",
                                      "ac|weights*", 'duration']

        self.add_subsystem('propmodel',
                           TwinSeriesHybridElectricPropulsionSystem(num_nodes=nn),
                           promotes_inputs=propulsion_promotes_inputs,
                           promotes_outputs=propulsion_promotes_outputs)
        self.connect('proprpm', ['propmodel.prop1.rpm', 'propmodel.prop2.rpm'])
        self.connect('hybrid_factor.vec', 'propmodel.hybrid_split.power_split_fraction')

        # use a different drag coefficient for takeoff versus cruise
        if flight_phase not in ['v0v1', 'v1v0', 'v1vr', 'rotate']:
            cd0_source = 'ac|aero|polar|CD0_cruise'
        else:
            cd0_source = 'ac|aero|polar|CD0_TO'
        self.add_subsystem('drag', PolarDrag(num_nodes=nn),
                           promotes_inputs=['fltcond|CL', 'ac|geom|*', ('CD0', cd0_source),
                                            'fltcond|q', ('e', 'ac|aero|polar|e')],
                           promotes_outputs=['drag'])

        self.add_subsystem('OEW',TwinSeriesHybridEmptyWeight(),
                           promotes_inputs=['*',('P_TO','ac|propulsion|engine|rating')],
                           promotes_outputs=['OEW'])
        self.connect('propmodel.propellers_weight', 'W_propeller')
        self.connect('propmodel.eng1.component_weight', 'W_engine')
        self.connect('propmodel.gen1.component_weight', 'W_generator')
        self.connect('propmodel.motors_weight', 'W_motors')
        intfuel = self.add_subsystem('intfuel', Integrator(num_nodes=nn, method='simpson', diff_units='s',
                                                              time_setup='duration'), promotes_inputs=['*'], promotes_outputs=['*'])
        intfuel.add_integrand('fuel_used', rate_name='fuel_flow', val=1.0, units='kg')
        self.add_subsystem('weight', AddSubtractComp(output_name='weight',
                                                     input_names=['ac|weights|MTOW', 'fuel_used'],
                                                     units='kg', vec_size=[1, nn],
                                                     scaling_factors=[1, -1]),
                           promotes_inputs=['*'],
                           promotes_outputs=['weight'])

class ElectricTwinAnalysisGroup(Group):
    """This is an example of a balanced field takeoff and three-phase mission analysis.
    """
    def setup(self):
        # Define number of analysis points to run pers mission segment
        nn = 11

        # Define a bunch of design varaiables and airplane-specific parameters
        dv_comp = self.add_subsystem('dv_comp',  DictIndepVarComp(acdata),
                                     promotes_outputs=["*"])
        dv_comp.add_output_from_dict('ac|aero|CLmax_TO')
        dv_comp.add_output_from_dict('ac|aero|polar|e')
        dv_comp.add_output_from_dict('ac|aero|polar|CD0_TO')
        dv_comp.add_output_from_dict('ac|aero|polar|CD0_cruise')

        dv_comp.add_output_from_dict('ac|geom|wing|S_ref')
        dv_comp.add_output_from_dict('ac|geom|wing|AR')
        dv_comp.add_output_from_dict('ac|geom|wing|c4sweep')
        dv_comp.add_output_from_dict('ac|geom|wing|taper')
        dv_comp.add_output_from_dict('ac|geom|wing|toverc')
        dv_comp.add_output_from_dict('ac|geom|hstab|S_ref')
        dv_comp.add_output_from_dict('ac|geom|hstab|c4_to_wing_c4')
        dv_comp.add_output_from_dict('ac|geom|vstab|S_ref')
        dv_comp.add_output_from_dict('ac|geom|fuselage|S_wet')
        dv_comp.add_output_from_dict('ac|geom|fuselage|width')
        dv_comp.add_output_from_dict('ac|geom|fuselage|length')
        dv_comp.add_output_from_dict('ac|geom|fuselage|height')
        dv_comp.add_output_from_dict('ac|geom|nosegear|length')
        dv_comp.add_output_from_dict('ac|geom|maingear|length')

        dv_comp.add_output_from_dict('ac|weights|MTOW')
        dv_comp.add_output_from_dict('ac|weights|W_fuel_max')
        dv_comp.add_output_from_dict('ac|weights|MLW')
        dv_comp.add_output_from_dict('ac|weights|W_battery')

        dv_comp.add_output_from_dict('ac|propulsion|engine|rating')
        dv_comp.add_output_from_dict('ac|propulsion|propeller|diameter')
        dv_comp.add_output_from_dict('ac|propulsion|generator|rating')
        dv_comp.add_output_from_dict('ac|propulsion|motor|rating')
        dv_comp.add_output('ac|propulsion|battery|specific_energy',val=300,units='W*h/kg')

        dv_comp.add_output_from_dict('ac|num_passengers_max')
        dv_comp.add_output_from_dict('ac|q_cruise')
        dv_comp.add_output_from_dict('ac|num_engines')

        mission_data_comp = self.add_subsystem('mission_data_comp',IndepVarComp(),promotes_outputs=["*"])
        mission_data_comp.add_output('batt_soc_target', val=0.1, units=None)

        analysis = self.add_subsystem('analysis',FullMissionAnalysis(num_nodes=nn,
                                                                     aircraft_model=SeriesHybridTwinModel),
                                                 promotes_inputs=['*'],promotes_outputs=['*'])

        margins = self.add_subsystem('margins',ExecComp('MTOW_margin = MTOW - OEW - total_fuel - W_battery - payload',
                                                        MTOW_margin={'units':'lbm','val':100},
                                                        MTOW={'units':'lbm','val':10000},
                                                        OEW={'units':'lbm','val':5000},
                                                        total_fuel={'units':'lbm','val':1000},
                                                        W_battery={'units':'lbm','val':1000},
                                                        payload={'units':'lbm','val':1000}),
                                                        promotes_inputs=['payload'])
        self.connect('cruise.OEW','margins.OEW')
        self.connect('descent.fuel_used_final','margins.total_fuel')
        self.connect('ac|weights|MTOW','margins.MTOW')
        self.connect('ac|weights|W_battery','margins.W_battery')

        augobj = self.add_subsystem('aug_obj', AugmentedFBObjective(), promotes_outputs=['mixed_objective'])
        self.connect('ac|weights|MTOW','aug_obj.ac|weights|MTOW')
        self.connect('descent.fuel_used_final','aug_obj.fuel_burn')

def configure_problem():
    prob = Problem()
    prob.model= ElectricTwinAnalysisGroup()
    prob.model.nonlinear_solver=NewtonSolver(iprint=1)
    prob.model.options['assembled_jac_type'] = 'csc'
    prob.model.linear_solver = DirectSolver(assemble_jac=True)
    prob.model.nonlinear_solver.options['solve_subsystems'] = True
    prob.model.nonlinear_solver.options['maxiter'] = 10
    prob.model.nonlinear_solver.options['atol'] = 1e-7
    prob.model.nonlinear_solver.options['rtol'] = 1e-7
    return prob

def set_values(prob, num_nodes, design_range, spec_energy): 
    # set some (required) mission parameters. Each pahse needs a vertical and air-speed
    # the entire mission needs a cruise altitude and range
    prob.set_val('climb.fltcond|vs', np.ones((num_nodes,))*1500, units='ft/min')
    prob.set_val('climb.fltcond|Ueas', np.ones((num_nodes,))*124, units='kn')
    prob.set_val('cruise.fltcond|vs', np.ones((num_nodes,))*0.01, units='ft/min')
    prob.set_val('cruise.fltcond|Ueas', np.ones((num_nodes,))*170, units='kn')
    prob.set_val('descent.fltcond|vs', np.ones((num_nodes,))*(-600), units='ft/min')
    prob.set_val('descent.fltcond|Ueas', np.ones((num_nodes,))*140, units='kn')

    prob.set_val('cruise|h0',29000,units='ft')
    prob.set_val('mission_range',design_range,units='NM')
    prob.set_val('payload',1000,units='lb')
    prob.set_val('ac|propulsion|battery|specific_energy', spec_energy, units='W*h/kg')

    # (optional) guesses for takeoff speeds may help with convergence
    prob.set_val('v0v1.fltcond|Utrue',np.ones((num_nodes))*50,units='kn')
    prob.set_val('v1vr.fltcond|Utrue',np.ones((num_nodes))*85,units='kn')
    prob.set_val('v1v0.fltcond|Utrue',np.ones((num_nodes))*85,units='kn')

    # set some airplane-specific values
    prob['analysis.cruise.acmodel.OEW.const.structural_fudge'] = 2.0
    prob['ac|propulsion|propeller|diameter'] = 2.2
    prob['ac|propulsion|engine|rating'] = 1117.2

def run_hybrid_twin_analysis(plots=False):
    prob = configure_problem()
    prob.setup(check=False)
    prob['cruise.hybridization'] = 0.05840626452293813
    set_values(prob, 11, 500, 450)
    prob.run_model()
    if plots:
        show_outputs(prob)
    return prob
    
def show_outputs(prob):
    # print some outputs
    vars_list = ['ac|weights|MTOW','climb.OEW','descent.fuel_used_final',
                 'rotate.range_final','descent.propmodel.batt1.SOC_final','cruise.hybridization',
                 'ac|weights|W_battery','margins.MTOW_margin',
                 'ac|propulsion|motor|rating','ac|propulsion|generator|rating','ac|propulsion|engine|rating',
                 'ac|geom|wing|S_ref','v0v1.Vstall_eas','v0v1.takeoff|vr',
                 'engineoutclimb.gamma']
    units = ['lb','lb','lb',
             'ft', None, None,
             'lb','lb',
             'hp','hp','hp',
             'ft**2','kn','kn',
             'deg']
    nice_print_names = ['MTOW', 'OEW', 'Fuel used',
                        'TOFL (over 35ft obstacle)', 'Final state of charge', 'Cruise hybridization',
                        'Battery weight','MTOW margin',
                        'Motor rating', 'Generator rating', 'Engine rating',
                        'Wing area', 'Stall speed', 'Rotate speed',
                        'Engine out climb angle']
    print("=======================================================================")
    for i, thing in enumerate(vars_list):
        print(nice_print_names[i]+': '+str(prob.get_val(thing,units=units[i])[0])+' '+str(units[i]))

    # plot some stuff
    plots = True
    if plots:
        x_var = 'range'
        x_unit = 'NM'
        y_vars = ['fltcond|h','fltcond|Ueas','fuel_used','throttle','fltcond|vs','propmodel.batt1.SOC']
        y_units = ['ft','kn','lbm',None,'ft/min',None]
        x_label = 'Range (nmi)'
        y_labels = ['Altitude (ft)', 'Veas airspeed (knots)', 'Fuel used (lb)', 'Throttle setting', 'Vertical speed (ft/min)', 'Battery SOC']
        phases = ['v0v1','v1vr','v1v0','rotate']
        plot_trajectory(prob, x_var, x_unit, y_vars, y_units, phases,
                        x_label=x_label, y_labels=y_labels, marker='-',
                        plot_title='Takeoff Profile')

        phases = ['v0v1','v1vr','v1v0','rotate','climb','cruise','descent']
        plot_trajectory(prob, x_var, x_unit, y_vars, y_units, phases,
                        x_label=x_label, y_labels=y_labels, marker='-',
                        plot_title='Full Mission Profile')

if __name__ == "__main__":
    # for run type choose choose optimization, comp_sizing, or analysis
    run_type = 'example'
    num_nodes = 11

    if run_type == 'example':
        # runs a default analysis-only mission (no optimization)
        run_hybrid_twin_analysis(plots=True)

    else:
        # can run a sweep of design range and spec energy (not tested)
        #design_ranges = [300,350,400,450,500,550,600,650,700]
        #specific_energies = [250,300,350,400,450,500,550,600,650,700,750,800]

        #or a single point
        design_ranges = [500]
        specific_energies = [450]

        write_logs = False
        if write_logs:
            logging.basicConfig(filename='opt.log', filemode='w', format='%(name)s - %(levelname)s - %(message)s')
        # run a sweep of cases at various specific energies and ranges
        for design_range in design_ranges:
            for this_spec_energy in specific_energies:
                try:
                    prob = configure_problem()
                    spec_energy = this_spec_energy
                    if run_type == 'optimization':
                        print('======Performing Multidisciplinary Design Optimization===========')
                        prob.model.add_design_var('ac|weights|MTOW', lower=4000, upper=5700)
                        prob.model.add_design_var('ac|geom|wing|S_ref',lower=15,upper=40)
                        prob.model.add_design_var('ac|propulsion|engine|rating',lower=1,upper=3000)
                        prob.model.add_design_var('ac|propulsion|motor|rating',lower=450,upper=3000)
                        prob.model.add_design_var('ac|propulsion|generator|rating',lower=1,upper=3000)
                        prob.model.add_design_var('ac|weights|W_battery',lower=20,upper=2250)
                        prob.model.add_design_var('ac|weights|W_fuel_max',lower=500,upper=3000)
                        prob.model.add_design_var('cruise.hybridization', lower=0.001, upper=0.999)
                        prob.model.add_design_var('climb.hybridization', lower=0.001, upper=0.999)
                        prob.model.add_design_var('descent.hybridization', lower=0.01, upper=1.0)

                        prob.model.add_constraint('margins.MTOW_margin',lower=0.0)
                        prob.model.add_constraint('rotate.range_final',upper=1357)
                        prob.model.add_constraint('v0v1.Vstall_eas',upper=42.0)
                        prob.model.add_constraint('descent.propmodel.batt1.SOC_final',lower=0.0)
                        prob.model.add_constraint('climb.throttle',upper=1.05*np.ones(num_nodes))
                        prob.model.add_constraint('climb.propmodel.eng1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('climb.propmodel.gen1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('climb.propmodel.batt1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('cruise.propmodel.eng1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('cruise.propmodel.gen1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('cruise.propmodel.batt1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('descent.propmodel.eng1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('descent.propmodel.gen1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('descent.propmodel.batt1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('v0v1.propmodel.batt1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('engineoutclimb.gamma',lower=0.02)
                        prob.model.add_objective('mixed_objective') # TODO add this objective

                    elif run_type == 'comp_sizing':
                        print('======Performing Component Sizing Optimization===========')
                        prob.model.add_design_var('ac|propulsion|engine|rating',lower=1,upper=3000)
                        prob.model.add_design_var('ac|propulsion|motor|rating',lower=1,upper=3000)
                        prob.model.add_design_var('ac|propulsion|generator|rating',lower=1,upper=3000)
                        prob.model.add_design_var('ac|weights|W_battery',lower=20,upper=2250)
                        prob.model.add_design_var('cruise.hybridization', lower=0.01, upper=0.5)

                        prob.model.add_constraint('margins.MTOW_margin',equals=0.0) # TODO implement
                        prob.model.add_constraint('rotate.range_final',upper=1357) # TODO check units
                        prob.model.add_constraint('descent.propmodel.batt1.SOC_final',lower=0.0)
                        prob.model.add_constraint('v0v1.propmodel.eng1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('v0v1.propmodel.gen1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('v0v1.propmodel.batt1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('climb.propmodel.eng1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('climb.propmodel.gen1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('climb.propmodel.batt1.component_sizing_margin',upper=1.0*np.ones(num_nodes))
                        prob.model.add_constraint('climb.throttle',upper=1.05*np.ones(num_nodes))
                        prob.model.add_objective('fuel_burn')

                    else:
                        print('======Analyzing Fuel Burn for Given Mision============')
                        prob.model.add_design_var('cruise.hybridization', lower=0.01, upper=0.5)
                        prob.model.add_constraint('descent.propmodel.batt1.SOC_final',lower=0.0)
                        prob.model.add_objective('descent.fuel_used_final')

                    prob.driver = ScipyOptimizeDriver()
                    if write_logs:
                        filename_to_save = 'case_'+str(spec_energy)+'_'+str(design_range)+'.sql'
                        if os.path.isfile(filename_to_save):
                            print('Skipping '+filename_to_save)
                            continue
                        recorder = SqliteRecorder(filename_to_save)
                        prob.driver.add_recorder(recorder)
                        prob.driver.recording_options['includes'] = []
                        prob.driver.recording_options['record_objectives'] = True
                        prob.driver.recording_options['record_constraints'] = True
                        prob.driver.recording_options['record_desvars'] = True

                    prob.setup(check=False)
                    set_values(prob, num_nodes, design_range, spec_energy)

                    run_flag = prob.run_driver()
                    if run_flag:
                        raise ValueError('Opt failed')

                except BaseException as e:
                    if write_logs:
                        logging.error('Optimization '+filename_to_save+' failed because '+repr(e))
                    prob.cleanup()
                    try:
                        os.rename(filename_to_save, filename_to_save.split('.sql')[0]+'_failed.sql')
                    except WindowsError as we:
                        if write_logs:
                            logging.error('Error renaming file: '+repr(we))
                        os.remove(filename_to_save)

        show_outputs(prob)